R/backend-file.R

```diff
@@ -0,0 +1,507 @@
+          TRUE
+        else
+          signalCondition(e)
+      }
+    )
+  }
+}
+
+b_file_keyring_default <- function(self, private) {
+  private$keyring
+}
+
+b_file_keyring_set_default <- function(self, private, keyring) {
+  private$keyring <- keyring
+  invisible(self)
+}
+
+## --------------------------------------------------------------------
+## Private
+
+b_file_keyring_create_direct <- function(self, private, keyring, password,
+  nonce, items) {
+
+  file_name <- keyring %||% private$keyring
+
+  if (file.exists(file_name))
+    confirmation(paste("are you sure you want to overwrite", file_name))
+
+  private$keyring_write_file(
+    file_name,
+    nonce %||% sodium::random(24L),
+    items %||% list(),
+    password %||% get_pass()
+  )
+
+  invisible(self)
+}
+
+b_file_keyring_file <- function(self, private, keyring, ...) {
+
+  file_name <- keyring %||% private$keyring
+
+  assert_that(is_string(file_name))
+
+  if (!file.exists(file_name))
+    private$keyring_create_direct(file_name, ...)
+
+  normalizePath(file_name, mustWork = TRUE)
+}
+
+b_file_read_keyring_file <- function(self, private, keyring) {
+
+  yml <- yaml::yaml.load_file(keyring %||% private$keyring)
+
+  assert_that(is.list(yml),
+              assertthat::has_name(yml, "keyring_info"),
+              is.list(yml[["keyring_info"]]),
+              assertthat::has_name(yml[["keyring_info"]], "nonce"),
+              assertthat::has_name(yml[["keyring_info"]], "integrity_check"),
+              assertthat::has_name(yml, "items"),
+              is.list(yml[["items"]]))
+
+  list(
+    nonce = sodium::hex2bin(yml[["keyring_info"]][["nonce"]]),
+    items = yml[["items"]],
+    check = yml[["keyring_info"]][["integrity_check"]]
+  )
+}
+
+b_file_write_keyring_file <- function(self, private, keyring, nonce, items,
+  key) {
+
+  nonce <- nonce %||% private$nonce_get(keyring)
+
+  yaml::write_yaml(
+    list(
+      keyring_info = list(
+        keyring_version = as.character(
+          utils::packageVersion(methods::getPackageName())
+        ),
+        nonce = sodium::bin2hex(nonce),
+        integrity_check = b_file_secret_encrypt(
+          paste(sample(letters, 22L, replace = TRUE), collapse = ""),
+          nonce,
+          key %||% private$key_get(keyring)
+        )
+      ),
+      items = items %||% private$items_get(keyring)
+    ),
+    keyring %||% private$keyring
+  )
+
+  invisible(self)
+}
+
+b_file_key_get <- function(self, private, keyring) {
+
+  kr_env <- b_file_keyring_env(private$keyring_file(keyring))
+
+  if (is.null(kr_env$key))
+    key <- private$key_set(keyring = keyring)
+  else
+    key <- kr_env$key
+
+  assert_that(is.raw(key), length(key) > 0L)
+
+  key
+}
+
+b_file_key_unset <- function(self, private, keyring) {
+
+  kr_env <- b_file_keyring_env(private$keyring_file(keyring))
+
+  kr_env$key <- NULL
+
+  invisible(kr_env)
+}
+
+b_file_key_is_set <- function(self, private, keyring)
```

Comment:
Please use `{` and `}`, everywhere, unless they would be on the same line.

Reply:
hopefully caught all (c786e4ababe9a19665b9aa1a6af8636f344723f0)